### PR TITLE
Maid 1437.

### PR DESCRIPTION
**Pull request by brian-js**
[Review on Reviewable](https://reviewable.io/reviews/maidsafe/routing/747)

### CODE REVIEW

**Comment by maidsafe-highfive** [comment]
r? @benjaminbollen

(maidsafe_highfive has picked a reviewer for you, use r? to override)

File: src/routing_node.rs [hunk 1]
```diff
@@ -690,36 +699,36 @@ impl RoutingNode {
...
-        };
-        match SignedMessage::new(Address::Client(core_id.signing_public_key()),
-                                 routing_message,
-                                 core_id.signing_private_key()) {
-            Ok(signed_message) => ignore(self.send(signed_message)),
-            Err(e) => return Err(RoutingError::Cbor(e)),
-        };
-        Ok(())
-    }
+    // fn request_network_name(&mut self,
```
Comment (benjaminbollen) [inline src/routing_node.rs]: request_network_name should be `used`

File: src/routing_core.rs [hunk 2]
```diff
@@ -743,9 +743,313 @@ impl RoutingCore {
...
+                            None => {},
+                        }
+                    }
+                }
+            },
+            (None, Some((unknown_connection, Some(hello)))) => {
+                // At matching from unknown_connection against expected connection
+                let mut opt_connection = None;
+                match hello.expected_connection {
+                    Some(ref hello_expected_connection) => {
```
Comment (benjaminbollen) [inline src/routing_core.rs]: ref and * can possibly be collapsed into single destructuring

File: src/routing_core.rs [hunk 3]
```diff
@@ -743,9 +743,313 @@ impl RoutingCore {
...
+                        match *hello_expected_connection {
+                            ExpectedConnection::Request(ref request) => {
+                                for (key, value) in self.expected_connections.iter() {
+                                    match key {
+                                        &ExpectedConnection::Request(_) => {
+                                            // Don't match on ConnectRequest.
```
Comment (benjaminbollen) [inline src/routing_core.rs]: I do believe we need to act here too.  We have a Hello and found a ConnectResponse

File: src/routing_core.rs [hunk 4]
```diff
@@ -743,9 +743,313 @@ impl RoutingCore {
...
+                                        },
+                                    }
+                                }
+                            },
+                        }
+                    },
+                    None => {},
+                };
+
+                match opt_connection {
```
Comment (benjaminbollen) [inline src/routing_core.rs]: make comment: this is the primary connection, either the expected connect_request connection for node B, or the unknown connection for node A - diagram RFC-0011

File: src/routing_core.rs [hunk 5]
```diff
@@ -743,9 +743,313 @@ impl RoutingCore {
...
+                    Some(connection) => {
```
Comment (benjaminbollen) [inline src/routing_core.rs]: but here we need to determine whether we are a client bootstrapping.

Alternatively, and as suggested by Brian, we need to earlier split the program flow and handle as a client before already

File: src/routing_core.rs [hunk 6]
```diff
@@ -743,9 +743,313 @@ impl RoutingCore {
...
+                        if self.add_peer(ConnectionName::Routing(
+                                hello.public_id.name()), connection, hello.public_id.clone()) {
+                            if connection != unknown_connection {
+                                self.action_sender.send(::action::Action::SendConfirmationHello(
+                                    connection, ::types::Address::Node(hello.public_id.name())));
+                                let _ = self.action_sender.send(::action::Action::DropConnections(
+                                    vec![unknown_connection]));
+                            } else {
```
Comment (benjaminbollen) [inline src/routing_core.rs]: the "confirmation of the hello" is not explicitly checked; which raises some questions, as if we already have a hello stored in unknown connections, then we drop any subsequent action on a hello.  So we need to check within handle_hello whether it is a consistent confirmation and do match.

File: src/routing_core.rs [hunk 7]
```diff
@@ -743,9 +743,313 @@ impl RoutingCore {
...
+
+    /// Check whether the connection has been accepted.
+    pub fn match_unknown_connection(&mut self, connection: &::crust::Connection,
+            hello: &::direct_messages::Hello) {
+        for (key, value) in self.unknown_connections.iter_mut() {
+            if key == connection {
+                match value.0 {
+                    None => {
+                        value.0 = Some(hello.clone());
+                        let _ = self.action_sender.send(::action::Action::MatchConnection(
```
Comment (benjaminbollen) [inline src/routing_core.rs]: Here we need to check whether it is a confirmation hello, consistent with the hello stored and continue to Match is so.

File: src/routing_node.rs [hunk 8]
```diff
@@ -293,142 +295,150 @@ impl RoutingNode {
...
-                // to avoid both sides drop the other connection, possibly leaving none
-                self.crust_service.drop_node(connection.clone());
-                debug!("Core refused {:?} on {:?} and dropped the connection",
-                    hello_address, connection);
-            };
-        } else {
-            debug!("We are not alpha and the hello was not confirmed yet, awaiting alpha.");
-        }
-        Ok(())
+        self.match_unknown_connection(&connection, &hello)
```
Comment (benjaminbollen) [inline src/routing_node.rs]: handle bootstrapping

**Comment by benjaminbollen** [comment]
looks good to me, without bootstrapping finished

**Pull request merged**